src/github.com/getlantern/flashlight/http/http.go

```diff
@@ -0,0 +1,167 @@
+package http
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/getlantern/flashlight/client"
+	"github.com/getlantern/flashlight/config"
+	"github.com/getlantern/flashlight/util"
+	"github.com/getlantern/golog"
+	"github.com/getlantern/whitelist"
+	"github.com/skratchdot/open-golang/open"
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/getlantern/tarfs"
+)
+
+const (
+	UiUrl = "http://%s"
+	UiDir = "src/github.com/getlantern/ui/app"
+)
+
+var (
+	log = golog.LoggerFor("http")
+)
+
+type JsonResponse struct {
+	Error     string   `json:"Error, omitempty"`
+	Whitelist []string `json:"Whitelist, omitempty"`
+	Original  []string `json:"Original, omitempty"`
+}
+
+type WhitelistHandler struct {
+	http.HandlerFunc
+	whitelist *whitelist.Whitelist
+	wlChan    chan *whitelist.Config
+}
+
+func sendJsonResponse(w http.ResponseWriter, response *JsonResponse, indent bool) {
+
+	enc := json.NewEncoder(w)
+	err := enc.Encode(response)
+	if err != nil {
+		log.Errorf("error sending json response %v", err)
+	}
+}
+
+func setResponseHeaders(w http.ResponseWriter) {
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+	w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token")
+	w.Header().Set("Access-Control-Allow-Credentials", "True")
+	w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
+
+}
+
+func (wlh WhitelistHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var response JsonResponse
+
+	setResponseHeaders(w)
+
+	switch r.Method {
+	case "OPTIONS":
+		// return if it's a preflight OPTIONS request
+		// this is mainly for testing the UI when it's running on
+		// a separate port
+		return
+	case "POST":
+		// update whitelist
+		decoder := json.NewDecoder(r.Body)
+		var entries []string
+		err := decoder.Decode(&entries)
+		if err != nil {
+			log.Error(err)
+			response.Error = fmt.Sprintf("Error decoding whitelist: %q", err)
+		} else {
+			wl := wlh.whitelist.UpdateEntries(entries)
+			copy := wlh.whitelist.Copy()
+			log.Debug("Propagating whitelist changes..")
+			wlh.wlChan <- copy
+			response.Whitelist = wl
+		}
+	case "GET":
+		response.Whitelist = wlh.whitelist.GetEntries()
+	default:
+		log.Debugf("Received %s", response.Error)
+		response.Error = "Invalid whitelist HTTP request"
+		response.Whitelist = nil
+	}
+	sendJsonResponse(w, &response, false)
+}
```

Comment:
@oxtoacart Yep, it does. I'm about finished with the websocket changes: should I create a second PR and merge into this branch? That'll actually be cleaner since there will be less files to inspect in the diff.